### PR TITLE
Add NestedSizeOfGcsObject

### PR DESCRIPTION
**Pull request by gargnitingoogle**
### Description
This adds implementation of NestedSizeOfGcsObject by
recursively calculating and adding up the nested sizes
of all its members.
Currently it's defined in a hard-coded fashion (i.e. members are
used through names rather than fetching them by using reflect package.)
Generalizing this is left as an improvement idea for later.

This is followed up in https://github.com/GoogleCloudPlatform/gcsfuse/pull/1681 .

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

**Comment by gargnitingoogle** [comment]
Unit tests need to be added, but otherwise code is ready for review.

File: internal/util/sizeof.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+	// integer-pointer members
+	size += unsafeSizeOf(o.CRC32C)
+	// pointer-to-integer-array members
+	size += unsafeSizeOf(o.MD5)
+
+	// integer members - Size, Generation, MetaGeneration, ComponentCount
+	// time members - Deleted, Updated
+	// boolean members - EventBasedHold
+	// nothing to be added for any built-in types - already accounted for in rawSizeOfValuePointedBy(o)
```
Comment (gargnitingoogle) [inline internal/util/sizeof.go:233]: Remove reference to obsolete function rawSizeOfValuePointedBy.

File: internal/util/sizeof.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	size += nestedContentSizeOfArrayOfStrings(&acl.ForceSendFields)
+	size += nestedContentSizeOfArrayOfStrings(&acl.NullFields)
+
+	return
+}
+
+// Assumes acls is not null, to avoid cost of checking.
+func nestedContentSizeOfArrayOfAclPointers(acls *[]*storage.ObjectAccessControl) (size int) {
+	for _, acl := range *acls {
+		// we could used unsafe.Sizeof(&acl) here instead of defining
```
Comment (gargnitingoogle) [inline internal/util/sizeof.go:195]: could -> could have

File: internal/util/sizeof.go (hunk 1 above)
Comment (gargnitingoogle) [reply internal/util/sizeof.go]: old obsolete comment.

**Comment by gargnitingoogle** [comment]
@sethiay please review this.

File: internal/util/sizeof.go [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+// Definitions/conventions (not based on a standard, but just made up for convenience).
+//   1. Raw/unsafe size: This is the size of empty
+//      data structures, i.e. size of memory, for data
+//      structures when just initialized, without any
+//      content filled into them.
+//      This is same as what is returned by unsafe.Sizeof(...)
+//
+//      Built-in types ([u]int*, bool, time.*) have the same
```
Comment (raj-prince) [inline internal/util/sizeof.go:42]: nit: same same -> same

File: internal/util/sizeof.go [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+//      same raw/unsafe size as their actual space in memory.
+//
+//      More complex types like string, array/slice have a fixed
+//      /standard raw/unsafe size e.g. string is 16 bytes (8 for
+//      holding pointer to its content, 8 bytes for holding length).
+//      But it doesn't account for string's content, so any string,
+//  	be it emoty-string, or "hello" or a million-char string, all
```
Comment (raj-prince) [inline internal/util/sizeof.go:49]: nit: empty-string

File: internal/util/sizeof.go [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+//      have Raw/unsafe size of 16.
```
Comment (raj-prince) [inline internal/util/sizeof.go:50]: nit: raw/unsafe

File: internal/util/sizeof.go [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+		return 0
+	}
+	return len(*s)
+}
+
+func nestedSizeOfString(s *string) int {
+	return UnsafeSizeOf(s) + contentSizeOfString(s)
+}
+
+func nestedContentSizeOfArrayOfStrings(arr *[]string) (size int) {
```
Comment (raj-prince) [inline internal/util/sizeof.go:108]: Can we stick with same terminology - contentSize or nestedContentSize?
I feel, contentSize will be okay, that means, size other than their type size.

File: internal/util/sizeof.go [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+		return
+	}
+
+	for _, str := range *arr {
+		size += nestedSizeOfString(&str)
+	}
+	return
+}
+
+func nestedSizeOfArrayOfStrings(arr *[]string) int {
```
Comment (raj-prince) [inline internal/util/sizeof.go:119]: shouldn't full size be more clear?

File: internal/util/sizeof.go [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+	}
+	return
+}
+
+func nestedContentSizeOfServerResponse(sr *googleapi.ServerResponse) (size int) {
+	if sr == nil {
+		return
+	}
+
+	// integer members - HTTPStatusCode
```
Comment (raj-prince) [inline internal/util/sizeof.go:152]: Why didn't we consider HTTPStatusCode?

Comment (gargnitingoogle) [reply internal/util/sizeof.go:152]: This is already explained in the documentation at the top in a general manner. There is no point in explaining it in every struct.

File: internal/util/sizeof.go (hunk 6 above)
Comment (gargnitingoogle) [reply internal/util/sizeof.go:108]: Switched all the contentSize for simplicity.

File: internal/util/sizeof.go [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+//  	be it emoty-string, or "hello" or a million-char string, all
+//      have Raw/unsafe size of 16.
+//
+//      Pointers have a raw/unsafe size of 8 bytes on 64-bit platforms.
+//
+//      Similarly, an array/slice/map etc. have small fixed Raw/unsafe size like 16 or 32 bytes.
+//
+//      Raw/unsafe size of a struct is the sum of Raw/unsafe sizes of all its constituents.
+//
+//   2. Content-size (or nested-content-size): This is the additional size added when some content is
```
Comment (raj-prince) [inline internal/util/sizeof.go:58]: I think, nestedSize and nestedContentSize both together hard to follow, specially when a new terminology content size comes into the picture.

We can chose only one and defined their behavior for struct, slices and maps should be enough.

Comment (gargnitingoogle) [reply internal/util/sizeof.go:58]: Not sure I follow the point. PLease clarify.

File: internal/util/sizeof.go [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+// of the gcs.Object pointed by the passed pointer.
+// Improvement scope: This can be generalized to a general-struct
+// but that needs better understanding of the reflect package
+// and other related packages.
+func NestedSizeOfGcsObject(o *gcs.Object) (size int) {
+	if o == nil {
+		return
+	}
+
+	// get raw size of the structure.
```
Comment (raj-prince) [inline internal/util/sizeof.go:241]: nit: First letter comment upper case.

File: internal/util/sizeof_test.go [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+			}{},
+			expected_size: sizeOfEmptyStruct + sizeOfInt,
+		},
+		{
+			t: struct {
+				a          int
+				b1, b2, b3 byte
+				c          string
+			}{},
+			expected_size: sizeOfEmptyStruct + sizeOfInt + 3*sizeOfByte + 5 /*for-padding-for-alignment*/ + sizeOfEmptyString,
```
Comment (raj-prince) [inline internal/util/sizeof_test.go:114]: I didn't get why 5?

File: internal/util/sizeof_test.go [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+			t:             map[string]int{"few ": 432, "fewfgwe": -21, "": 1, "fewawef": 0},
+			expected_size: sizeOfEmptyStringIntMap,
+		},
+	} {
+		calculatedSize := UnsafeSizeOf(&tc.t)
+		AssertEq(tc.expected_size, calculatedSize)
+	}
+}
+
+func (t *SizeofTest) TestContentSizeOfString() {
```
Comment (raj-prince) [inline internal/util/sizeof_test.go:142]: Since, we are testing both methods, contentSizeOfString and nestedSizeOfString.

So, we can rename:
Test_ContentSizeOfString_And_NestedCotentSizeOfString() {
}

and table struct will contain three entries - string, expected_content_size and expected_nested_content_size.

File: internal/util/sizeof_test.go [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+		{
+			m:                            map[string]string{"a": "b1", "xyz": "alpha"},
+			expected_nested_content_size: sizeOfEmptyString + 1 + sizeOfEmptyString + 2 + sizeOfEmptyString + 3 + sizeOfEmptyString + 5,
+		},
+	} {
+		AssertEq(tc.expected_nested_content_size, nestedContentSizeOfStringToStringMap(&tc.m))
+	}
+}
+
+func (t *SizeofTest) TestNestedContentSizeOfStringToStringArrayMap() {
```
Comment (raj-prince) [inline internal/util/sizeof_test.go:226]: We should take a case, whose value is an array of string of size 2 or more than 2.

File: internal/util/sizeof_test.go [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+	expectedSize += len(contentType) + len(name) + len(contentLanguage) + len(cacheControl) + len(contentEncoding) + len(owner) + len(mediaLink) + len(storageClass) + len(contentDisposition) + len(customTime)
+	expectedSize += md5.Size // for MD5 [md5.Size]byte
+	expectedSize += 4        // for CRC 32 uint32
+	expectedSize += customMetadataFieldsContentSize
+	expectedSize += nestedContentSizeOfArrayOfAclPointers(&customAcls)
+
+	AssertEq(expectedSize, NestedSizeOfGcsObject(&o))
+}
+
+func (t *SizeofTest) TestNestedSizeOfMinObject() {
```
Comment (raj-prince) [inline internal/util/sizeof_test.go:372]: We can create a similar test as we did for GCSObject.

**Review by raj-prince — CHANGES_REQUESTED** [review]
Logically looks good to me. A couple of minor comments.

File: internal/util/sizeof_test.go (hunk 14 above)
Comment (gargnitingoogle) [reply internal/util/sizeof_test.go:372]: As NestedSizeOfMinObject isn't used/needed anywhere today, I've removed it altogether. When we switch MinObject, we can switch NestedSizeOfGcsObject itself to NestedSizeOfMinObject.

File: internal/util/sizeof.go (hunk 10 above)
Comment (gargnitingoogle) [reply internal/util/sizeof.go:241]: done.

File: internal/util/sizeof_test.go (hunk 13 above)
Comment (gargnitingoogle) [reply internal/util/sizeof_test.go:226]: Done. Updated the last test case.

File: internal/util/sizeof_test.go (hunk 12 above)
Comment (gargnitingoogle) [reply internal/util/sizeof_test.go:142]: Renamed. The reason I didn't keep expected_nested_size separately as it's always expected to be = sizeOfEmptyString + expected_content_size, so duplication of code. Anyway, I've added it now.

File: internal/util/sizeof_test.go [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+			str:                   "hello-world",
+			expected_content_size: 11,
+		},
+	} {
+		AssertEq(tc.expected_content_size, contentSizeOfString(&tc.str))
+		AssertEq(tc.expected_content_size+sizeOfEmptyString, nestedSizeOfString(&tc.str))
+	}
+}
+
+func (t *SizeofTest) TestNestedContentSizeOfArrayOfStrings() {
```
[the remark this answers is not in the document]
Comment (gargnitingoogle) [reply internal/util/sizeof_test.go:165]: done here also.

File: internal/util/sizeof_test.go (hunk 11 above)
Comment (gargnitingoogle) [reply internal/util/sizeof_test.go:114]: 5 is what compiler pads between b3 and c to align c to a multiple of 8 (a is at offset 0, b1 at 8, b2 at 9, b3 at 10, c at 16). This is a compiler feature. Compilers always need to align structures within structures at offsets which are multiples of the platform-int-size (8 bytes in this case).

File: internal/util/sizeof.go (hunk 9 above)
Comment (gargnitingoogle) [reply internal/util/sizeof.go:58]: @raj-prince please respond.

File: internal/util/sizeof.go (hunk 7 above)
Comment (raj-prince) [reply internal/util/sizeof.go:123]: We can skip then. Nested will make sense, since we renamed nestedContent -> Content.

File: internal/util/sizeof.go (hunk 9 above)
Comment (raj-prince) [reply internal/util/sizeof.go:58]: I was saying, content and nested-content creates confusion. Hence, only content-size will be okay. As you have already defined what does content size means for structs, slices and maps.

Comment (gargnitingoogle) [reply internal/util/sizeof.go:58]: Removed the name nested-content-size.

**Comment by gargnitingoogle** [comment]
@raj-prince resolved all comments and did some clean-up
